src/Interpreters/TreeOptimizer.cpp

```diff
@@ -172,16 +172,29 @@ void optimizeGroupBy(ASTSelectQuery * select_query, ContextPtr context)
             /// copy shared pointer to args in order to ensure lifetime
             auto args_ast = function->arguments;
 
-            /** remove function call and take a step back to ensure
-              * next iteration does not skip not yet processed data
-              */
-            remove_expr_at_index(i);
-
-            /// copy non-literal arguments
-            std::remove_copy_if(
-                    std::begin(args_ast->children), std::end(args_ast->children),
-                    std::back_inserter(group_exprs), is_literal
-            );
+            if (select_query->group_by_with_rollup)
```

Comment:
let's do it this way in all cases?

Reply:
Thank you for your review. `std::swap()` + `remove_copy_if()` only places elements at the end of the array, which is relatively more efficient. On the other hand, `std::copy_if()` + `std::inserter()`, which involves inserting at a specific position in the array, could lead to multiple element movements, especially when `group_exprs` is very large. Therefore, I have retained the use of `std::swap()` +` remove_copy_if()`, and only use `std::inserter()` in the case of `group_by_with_rollup`.

Reply:
it makes sense. though I think we could make the code simpler (by just having less if-s). performance in this place prob not a concern since I'd say more than 10 group by keys is already a marginal case.
also it might be that you'll occasionally fix some other issues. e.g. it seems possible that `with cube` has the same problem. could you pls check it btw?

Reply:
You're right, we should make the code simpler. The case with a large number of group by keys is also a marginal case; I'll make the changes. Additionally, let me also check the scenario involving `with cube`.

Reply:
@nickitat Hello, I have revised a simpler version of the code. Additionally, I looked into the logic of `with cube`. Since `with cube` generates all possible grouping combinations, the order of `group by keys` does not affect the `with cube` query, so there's no need to worry about `with cube` having this bug.